### PR TITLE
Gradient colorscale nodata

### LINKED ISSUE #120: Backend: gradient colormap and no data?
Mbio does not want the no data option, so for first implementations of the gradient colormap in alpha div only we don't need to worry about it. However, when it gets introduced to the pass app scatter, we'll need a plan.

If we let no data be on with the gradient colormap, that technically works because the no data are different markers. However, it makes handling the legend tricky because it will need both a gradient legend and a row for no data. I'm constructing the gradient colormap legend to be a component that i hope could be used by our custom legend so that this option would be possible. We'll also need some plot.data updates.

If we disallow no data + gradient colorscale we'll need to just add another rule for disabling the toggle.

I'm leaning towards implementing the second option until someone tells us that we definitely need no data with gradient colormap.

### PR DESCRIPTION
**Pull request by asizemore**
Resolves #120 

See also data service [PR #138](https://github.com/VEuPathDB/EdaDataService/pull/138), web-eda [PR #790](https://github.com/VEuPathDB/web-eda/pull/790), and web-components [PR #275](https://github.com/VEuPathDB/web-components/pull/275).

When `evilMode=T` and the overlay variable is continuous, any group with missing overlay data gets a new row in which the `seriesGradientColormap` is 'NA' for all (x,y) points in that row. This is similar to how no data is handled for categorical overlays, but instead of an explicit value/no data label, the relevant column returns values or 'NA'. The client will then use the 'NA' to render that point with the proper marker and color.

Still up for debate is var constraints and colorscales. Leaving as dataShape=continuous for now.

Unit tests added and passing. Waiting for integration test before asking for review.

### CODE REVIEW

**Comment by asizemore** [comment]
If you squint really hard you can see tiny little 'x's! Marking for review

Screen Shot 2022-01-05 at 11 48 36 AM

File: tests/testthat/test-scattergl.R [hunk 1]
```diff
@@ -655,4 +655,28 @@ test_that("scattergl.dt() returns correct information about missing data", {
...
+  map <- data.frame('id' = c('entity.contC', 'entity.contB', 'entity.contA'),
+                  'plotRef' = c('overlayVariable', 'yAxisVariable', 'xAxisVariable'),
+                  'dataType' = c('NUMBER', 'NUMBER', 'NUMBER'),
+                  'dataShape' = c('CONTINUOUS', 'CONTINUOUS', 'CONTINUOUS'), stringsAsFactors=FALSE)
+
+  dt <- scattergl.dt(df, map, 'raw', evilMode = T)
+  expect_equal(nrow(dt), 2)
+  expect_equal(names(dt), c('seriesX', 'seriesY', 'seriesGradientColorscale'))
+  expect_equal(lapply(dt$seriesGradientColorscale, length), lapply(dt$seriesX, length))
+  expect_true(all(is.na(dt$seriesGradientColorscale[[2]])))
```
[the remark this answers is not in the document]
Comment (asizemore) [reply tests/testthat/test-scattergl.R:671]: Yup that's perfect! We can't test the NULL exactly (see [issue 70](https://github.com/jeroen/jsonlite/issues/70)) when reading in with `fromJSON` but we can test that NAs go in and that `fromJSON` gives us NAs. I manually validated that NAs are becoming nulls in the output json file written when calling `scattergl`.

File: R/class-plotdata.R [hunk 2]
```diff
@@ -201,7 +201,7 @@ newPlotdata <- function(.dt = data.table(),
   veupathUtils::logWithTime('Determined total number of complete cases across axes and strata vars.', verbose)
 
   if (evilMode) {
-    if (!is.null(group)) { .dt[[group]][is.na(.dt[[group]])] <- 'No data' }
+    if (!is.null(group) && !is.numeric(.dt[[group]])) { .dt[[group]][is.na(.dt[[group]])] <- 'No data' }
```
[the remark this answers is not in the document]
Comment (asizemore) [reply R/class-plotdata.R]: Comment added and code clarified! Lmk if the update makes more sense!

File: R/utils-json.R [hunk 3]
```diff
@@ -130,15 +130,12 @@ getJSON <- function(.pd, evilMode) {
...
-  #   namedAttrList$collectionVariableDetails <- makeVariableDetails(NULL, namedAttrList$collectionVariable$variableId, namedAttrList$collectionVariable$entityId, namedAttrList$collectionVariable$displayLabel)
-  #   namedAttrList$collectionVariable <- NULL
-  # }
   
   .pd <- addStrataVariableDetails(.pd)
   # If overlay is continuous, handle similarly to x, y, z vars.
-  if ('overlayVariable' %in% names(namedAttrList) & identical(namedAttrList$overlayVariable$dataShape, 'CONTINUOUS')) {
+  if ('overlayVariable' %in% names(namedAttrList) && identical(namedAttrList$overlayVariable$dataShape, 'CONTINUOUS')) {
     namedAttrList$overlayVariableDetails <- makeVariableDetails(NULL, namedAttrList$overlayVariable$variableId, namedAttrList$overlayVariable$entityId, namedAttrList$overlayVariable$displayLabel)
+    .pd$seriesGradientColorscale <- lapply(.pd$seriesGradientColorscale, function(x) {ifelse(is.na(x), 'NA', x)})
```
[the remark this answers is not in the document]
Comment (asizemore) [reply R/utils-json.R]: removed because of the lineplot update

**Review by d-callan — APPROVED** [review]

**Pull request merged**